### PR TITLE
Zipped NodeJS modules as actions

### PR DESCRIPTION
**Pull request by psuter**
This PR introduces support for Node actions written as zipped modules.

Builds on #1348 which should be merged first.

### CODE REVIEW

**Comment by psuter** [comment]
PG2 381 is blue-ish

**Comment by bjustin-ibm** [comment]
It might be nice to supply a [Yeoman generator](http://yeoman.io/learning/index.html) to help generate OpenWhisk actions.

**Comment by bjustin-ibm** [comment]
Thinking longer term, it might be advisable to support .tgz files produced with the `npm pack` command. `npm pack` is smart enough to include "bundled" dependencies and exclude files that are not needed for production use (automated tests, build files, README, etc.).

This would require that `npm install` is run on the .tgz each time, however, if all the required dependencies are bundled, this command is essentially equivalent to just unpacking the .tgz. However, if all the required dependencies are not bundled, the installation will still work (albeit slower due to the overhead of fetching dependencies). At this point, using `bundledDependencies` is an optimization we can strongly encourage to the user through documentation, the aforementioned Yeoman generator, and even log messages when running actions that did not include all the required deps.

File: core/nodejsActionBase/runner.js [hunk 1]
```diff
@@ -108,6 +127,55 @@ function NodeActionRunner(whisk) {
...
+    // decompress it into temporary directory, and return the name of that directory.
+    // Note that this makes heavy use of shell commands because:
+    //   1) Node 0.12 doesn't have many of the useful fs functions.
+    //   2) We know in which environment we're running.
+    function unzipInTmpDir(base64) {
+        var mkTempCmd = "mktemp -d XXXXXXXX";
+        return exec(mkTempCmd).then(function (tmpDir1) {
+            return new Promise(
+                function (resolve, reject) {
+                    var zipFile = tmpDir1 + "/action.zip";
```
Comment (bjustin-ibm) [inline core/nodejsActionBase/runner.js]: Using the `path` module to join paths is ever-so-slightly nicer:
```javascript
var zipFile = path.join(tmpDir1, 'action.zip')
```

File: core/nodejsActionBase/runner.js [hunk 2]
```diff
@@ -47,26 +50,42 @@ function NodeActionRunner(whisk) {
...
+                return Promise.reject("Could not uncompress action archive.");
+            });
+        } else {
+            // The code is a plain old JS file.
+            try {
+                eval(message.code);
+                thisRunner.userScriptMain = eval(message.main);
+                assertMainIsFunction()
+                // See comment above about 'true'; it has no specific meaning.
+                return Promise.resolve(true);
```
Comment (bjustin-ibm) [inline core/nodejsActionBase/runner.js]: 😒

File: core/nodejsActionBase/runner.js [hunk 3]
```diff
@@ -47,26 +50,42 @@ function NodeActionRunner(whisk) {
...
+            } catch (e) {
+                return Promise.reject(e);
```
Comment (psuter) [inline core/nodejsActionBase/runner.js]: I did not know that! :smiley: Thanks for the education.

File: core/nodejsActionBase/runner.js [hunk 4]
```diff
@@ -108,6 +127,55 @@ function NodeActionRunner(whisk) {
...
+                });
+            }
+        );
+    }
+
+    // Not exactly correct as it doesn't check the number of ='s at the end.
+    var base64Regex = /^([0-9a-zA-Z+/])*=*$/;
+    function isBase64(str) {
+        return base64Regex.test(str);
+    }
```
Comment (psuter) [inline core/nodejsActionBase/runner.js]: Sounds good.

File: core/nodejsActionBase/runner.js [hunk 5]
```diff
@@ -21,6 +21,9 @@
  * This file (runner.js) must currently live in root directory for nodeJsAction.
  */
 var util = require('util');
+var child_process = require('child_process');
+var fs = require('fs');
+var path = require('path');
```
Comment (psuter) [inline core/nodejsActionBase/runner.js]: Out of curiosity, why is that better? Is it a loading time issue?

Comment (bjustin-ibm) [inline core/nodejsActionBase/runner.js]: Exactly. `require` immediately goes out to disk reads the file and sticks it in memory. Lazy loading modules saves both IO time and memory usage in paths where you don't call the `require`.

Comment (psuter) [inline core/nodejsActionBase/runner.js]: It's a delicate balance then.. Assuming a warm js container, `require`ing at the top saves some latency when initializing a zip-based action. Assuming a cold container, it wastes some time for non-zip actions?

File: tests/src/actionContainers/NodeJsActionContainerTests.scala [hunk 6]
```diff
@@ -434,4 +436,60 @@ class NodeJsActionContainerTests extends BasicActionRunnerTests with WskActorSys
...
+            c.init(initPayload(code))._1 should not be (200)
+        }
+
+        // Somewhere, the logs should mention the connection to the archive.
+        checkStreams(out, err, {
+            case (o, e) =>
+                (o + e).toLowerCase should include("error")
+                (o + e).toLowerCase should include("archive")
+        })
+    }
```
Comment (psuter) [inline tests/src/actionContainers/NodeJsActionContainerTests.scala]: Yes

**Comment by psuter** [comment]
The way I see it, supporting `.tgz` could be part of this PR, but supporting `npm install` seems beyond its scope.

File: core/nodejsActionBase/runner.js (hunk 5 above)
Comment (bjustin-ibm) [inline core/nodejsActionBase/runner.js]: I believe what you are saying is correct. My gut feeling is that in all cases, we should delay this loading until we know we need it. If that initialization happens to get re-used by a warm container, then that is even better.

**Comment by bjustin-ibm** [comment]
> The way I see it, supporting .tgz could be part of this PR, but supporting npm install seems beyond its scope.

I agree, and I even think that supporting `.tgz` without `npm install` may be more confusing than valuable. In this way, we may be able to say `.zip` is for completely self-contained archives, and `.tgz` is for using with `npm install`.

**Comment by psuter** [comment]
Addressed the `binary` marker (#1391 needs to be merged first).

Improved the error messages for the various failure modes of uploading a broken/incomplete zip file.

Added tests for the above.

**Comment by psuter** [comment]
PG2 415

**Comment by rabbah** [comment]
Merged as https://github.com/openwhisk/openwhisk/compare/cb02a8850a4e...b36fd6faafbe.

**Pull request closed without merging**